src/Npgsql/Internal/NpgsqlConnector.cs

```diff
@@ -491,6 +493,18 @@ internal async Task Open(NpgsqlTimeout timeout, bool async, CancellationToken ca
                 OpenTimestamp = DateTime.UtcNow;
                 Log.Trace($"Opened connection to {Host}:{Port}");
 
+                if (PhysicalOpenCallback is not null)
+                {
+                    var physicalOpenTask = PhysicalOpenCallback(this);
+                    if (!async && !physicalOpenTask.IsCompleted)
+                    {
+                        physicalOpenTask.GetAwaiter().GetResult();
+                        throw new InvalidOperationException($"Asynchronous {nameof(PhysicalOpenCallback)} isn't supported with synchronous {nameof(NpgsqlConnection.Open)}");
+                    }
+
+                    await physicalOpenTask;
```

Comment:
Small nit: if we're in sync mode, we still await the task - this is OK since we've checked above that it's completed. But we could just skip that altogether. So, if the task is complete, no need to do anything (in either sync or async), if it's incomplete we await it in async and throw in sync.

Reply:
We actually should for ValueTask ;)

Reply:
That's true, but it's a Task now :)

Another reason why I prefer using Task rather than ValueTask everywhere by default. The contract for Task is much simpler and less bug-prone, and when there's really no perf justification for it...

Reply:
Ah I hadn't seen that change, yes for task it can really just short circuit!

Reply:
Totally off-topic, but I never understood why they chose to integrate IValueTaskSource into ValueTask as opposed to create a new awaitable. That would have made it typingly obvious what a method returns and what you're allowed to do to it. The non-IValueTaskSource ValueTask (which either contains the value or wraps a Task) is pretty simple and can also be reused without danger, but you can never be sure what's actually behind a ValueTask...

Reply:
Bit of the issue with the .NET type system is you have no nice existential return type support (without lifting it visibly into the type as a universal type). So you can either return a concrete struct or always pay for an interface. GetAwaiter for task already returned the concrete TaskAwaitable struct so I don't see they had much choice.

If you want to be sure you can do whatever with a valuetask you can always call Preserve, and it's nicely short circuited when it already wraps a task.

Reply:
But I mean, why could they not have created a public ReusableValueTask (or whatever name), which is the awaitable wrapping IValueTaskSource, instead of folding that functionality into the existing ValueTask, whose purpose at that time was only to optimize synchronous completion?

Reply:
I know Stephen Toub really doesn't like having too many user visible types, and I can imagine ValueTask & Task is already concept heavy for many users.